### DISCUSSION TITLE
Observable WeakMap and WeakSet

### DISCUSSION THREAD

**Discussion by upsuper** [mobxjs/mobx#3904] [Ideas]
In Canva, we have been finding it useful to have observable weak map, that is the observable equivalent of JavaScript [`WeakMap`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap). This is useful because we can put an object as key there and never worry about the object would leak from the map, while keeping its content observable in a similar way as observable map (that adding / deleting / changing values can trigger reactions that is observing the given key).

It is different from an observable map with `onBecomeUnobserved` or `createTransformer` in the sense that it doesn't require anything to observe it in order to keep the value around.

This would provide more parity with native JavaScript data structures, but I can also see reasons we may not want it in MobX:
* it is straightforward to implement with existing MobX API, and
* it would indeed come at a maintenance cost and size bloat for a possibly less used feature.

Given these, it would be good to know whether the maintainers of MobX are interested in having observable weak map (and possibly also observable weak set) into MobX. Internally, we already implemented it using an additional `WeakMap<K, IAtom>` which is very straightforward with the exactly same interface as `WeakMap`. But I can see it is more involved for it to be implemented as part of MobX with all the additional features, so even though I could contribute an implementation, I wouldn't put too much effort into it without consent from project maintainers.

**Comment by mweststrate** [comment]
No strong opinion on it yet :) Could you elaborate a bit more on the semantics here. I imagine:

1. any observer to a value would prevent the item from being removed from the weakmap (probably intended)
2. how would you observe the keys of the map? E.g. I imagine that a GC of a value might not trigger an update to the keys(), so that observing the keys in practice might get out of sync with what is actually in the maybe?
3. Is there a practical reason (as far as you can tell now) that this should ideally be in core MobX and not in say, mobx-utils?

**Reply by upsuper** [reply]
> any observer to a value would prevent the item from being removed from the weakmap (probably intended)

If there is an observer that observes the item through this map, then they must be holding a strong reference to the key, and thus the item wouldn't be removed from the weak map regardless. I don't think one should be able to observe a weak map value without going through the key.

> how would you observe the keys of the map?

You just can't. JavaScript `WeakMap` doesn't provide an API to iterate through it for good reason. If you don't have the key, you can't access the value mapping from the key. This is a basic contract of `WeakMap`.

> Is there a practical reason (as far as you can tell now) that this should ideally be in core MobX and not in say, mobx-utils?

The main reason would be for parity between native data structures and observable data structures. Specifically, you may not be able to convert a native one into an observable one without it being in core MobX, for example by using decorator like `@mobx.observable.shallow` on a `WeakMap` or creating deep observing of a value with `WeakMap` inside.

**Reply by mweststrate** [reply]
> You just can't. JavaScript WeakMap doesn't provide an API to iterate through it for good reason. If you don't have the key, you can't access the value mapping from the key. This is a basic contract of WeakMap.

Ugh obviously. Shouldn't be responding to comments late at night when my brain already gave up 😂

> Specifically, you may not be able to convert a native one into an observable one without it being in core MobX, for example by using decorator 

Ok, that makes sense. Feel free to put up a rough draft PR to explore the further ramifications are!
